### PR TITLE
Created global thread pool object that classes can call when another thread is needed

### PR DESCRIPTION
**Pull request by altair1673**
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Whenever a class needs a new thread, a new thread pool was being created. Now there are global thread pool objects that can be called when one or more threads are needed. This reduces overhead since creating a new thread pool every time is not efficient.
### Why Should This Be In Core?

Potentially boost the speed of certain aspects of the app now that there is a shared thread pool now.
Cleaner code
### Benefits

Make things faster since overhead is reduced. (This still needs to be tested so we know exactly which components are faster).
### Possible Drawbacks

It was mentioned that some components might need their own thread pools but so far everything looks good.
The help modules doesn't use the new thread pool class since there is a cyclic dependency between it and the utilities module.
### Verification Process

Speed still needs to be tested.
But components seem to be working fine.

### Applicable Issues

#1504

### CODE REVIEW

**Review by Delphinus8821 — APPROVED** [review]

File: CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/templates/WorkflowQueryPlugin.java [hunk 1]
```diff
@@ -91,7 +92,9 @@ protected void execute(final PluginGraphs pluginGraphs, final PluginInteraction
 
         // create a service for executing jobs, limiting concurrent executions to the max concurrent plugins parameter.
         final int maxConcurrentPlugins = parameters.getIntegerValue(MAX_CONCURRENT_PLUGINS_PARAMETER_ID);
-        final ExecutorService workflowExecutor = Executors.newFixedThreadPool(maxConcurrentPlugins);
+
+        // THERE IS A POSSIBLE BUG HERE, GO BACK TO OLD METHOD IF ANY ISSUES
```
Comment (OrionsGuardian) [inline CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/templates/WorkflowQueryPlugin.java:96]: POSSIBLE BUG? .... Is this comment still appropriate?

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * This class holds global objects for the variouse thead pools needed by the
+ * application
+ *
+ * @author altair1673
+ */
+public class UniversalThreadPool {
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java:28]: Can we rename this to something like `ConstellationGlobalThreadPool`? Universal doesn't seem quite like the correct wording here (as it really is only for Constellation code use)

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * limitations under the License.
+ */
+package au.gov.asd.tac.constellation.utilities.threadpool;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * This class holds global objects for the variouse thead pools needed by the
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java:23]: "various thread"

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+public class UniversalThreadPool {
+
+    private static UniversalThreadPool threadPool = null;
+    private ScheduledExecutorService scheduledExecutorService = null;
+    private ExecutorService fixedThreadPool = null;
+    private ExecutorService cachedThreadPool = null;
+    private ExecutorService defaultPluginEnvironmentPool = null;
+
+    private UniversalThreadPool() {
+
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java:37]: can you remove the blank line here

**Review by antares1470 — CHANGES_REQUESTED** [review]
Are you possibly able to add an entry for this to the changelog? This would probably be a good change for developers to be aware of

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java (hunk 2 above)
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/threadpool/UniversalThreadPool.java:28]: One other thing I missed before, are you able to write some unit tests for this class?

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ */
+package au.gov.asd.tac.constellation.utilities.threadpool;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.testfx.api.FxToolkit;
+import static org.testng.Assert.*;
```
Comment (antares1470) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java:24]: are you able to change this import to only include the classes actually being used?

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * Test of getThreadPool method, of class ConstellationGlobalThreadPool.
+     */
+    @Test
+    public void testGetThreadPool() {
+        System.out.println("getThreadPool");
+        assertEquals(ConstellationGlobalThreadPool.getThreadPool() instanceof ConstellationGlobalThreadPool, true);
+    }
```
Comment (antares1470) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java:73]: Is there a way you can add to this test (and others) to demonstrate that the object you get back is what you're expecting (i.e. some asserts about the stuff in the object itself)? e.g. Perhaps demonstrating that you get the same object back if you call it a second time?

Acknowledge that there isn't much to those functions that you can really test but at present, the test is only really checking that what you returned matched the return type of the function...which doesn't really demonstrate anything about the function (if the return type didn't match, you'd get compilation errors)

**Review by antares1470 — CHANGES_REQUESTED** [review]
If addition to the comments, it looks like CI is failing, possibly some other places you need to edit in light of changes made after my comments

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java [hunk 7]
```diff
@@ -78,7 +80,9 @@ public void testGetThreadPool() {
     @Test
     public void testGetScheduledExecutorService() {
         System.out.println("getScheduledExecutorService");
-        assertEquals(ConstellationGlobalThreadPool.getThreadPool().getScheduledExecutorService() instanceof ScheduledExecutorService, true);
+        ScheduledExecutorService s1 = ConstellationGlobalThreadPool.getThreadPool().getScheduledExecutorService();
+        ScheduledExecutorService s2 = ConstellationGlobalThreadPool.getThreadPool().getScheduledExecutorService();
+        assertEquals(s1 == s2, true);
```
Comment (antares1470) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/threadpool/ConstellationGlobalThreadPoolNGTest.java:85]: Might be worth adding a comment explaining that you're checking that a new instance isn't being created the 2nd time the function is called (based on the function implementation). Similar thing for the other tests

**Review by antares1470 — APPROVED** [review]

**Pull request merged**